test/functional/feature_smart_contract_test.py

```diff
@@ -56,97 +73,290 @@ def setup_network(self):
 
-        # Find a suitable UTXO
         initial_utxo = [x for x in client.utxos_for(alice) if x[1].value >= 50][0]
+        value = initial_utxo[1].json()["value"]
+
+        self.log.error(initial_utxo)
 
-        tx0 = utxo.Transaction(
+        tx = utxo.Transaction(
             client,
             inputs=[
                 utxo.Input(initial_utxo[0]),
             ],
             outputs=[
                 utxo.Output(
-                    value=50,
+                    value=value,
                     destination=utxo.DestPubkey(alice.public_key),
-                    data=None
+                    data=None,
                 ),
-                utxo.Output(
-                    value=10,
-                    destination=utxo.DestCreatePP(
-                        code=os.path.join(os.path.dirname(__file__), "code.wasm"),
-                        data=[0xed, 0x4b, 0x9d, 0x1b],  # default() constructor selector
-                    ),
-                    data=None
-                ),
-                # This output prevent reward overflow
-                utxo.Output(
-                    value=3981553255926290448385, # = genesis amount - u64::MAX
-                    destination=utxo.DestPubkey(alice.public_key),
-                    data=None
-                )
             ]
         ).sign(alice, [initial_utxo[1]])
+        client.submit(alice, tx)
 
-        # submit transaction and get the extrinsic and block hashes
-        (ext, blk,_) = client.submit(alice, tx0)
+        # invalid bytecode
+        value -= 1
 
-        # each new smart contract instantiation creates a new account
-        # fetch this SS58-formatted account address and return it
-        # and the hex-encoded account id
-        (ss58, acc_id) = contract.getContractAddresses(substrate, blk)
+        (tx, (_, blk, _)) = submit_pp_tx(client, tx, alice, value, utxo.Output(
+            value=1,
+            destination=utxo.DestCreatePP(
+                code=[0x00],
+                data=[0xed, 0x4b, 0x9d, 0x1b],
+            ),
+            data=None,
+        ))
+        assert_equal(contract.getContractAddresses(substrate, blk), None)
+
+        # invalid value
+        (invalid_tx, res) = submit_pp_tx(client, tx, alice, value, utxo.Output(
+            value=0,
+            destination=utxo.DestCreatePP(
+                code=os.path.join(os.path.dirname(__file__), "assets/pooltester.wasm"),
+                data=[0xed, 0x4b, 0x9d, 0x1b],
+            ),
+            data=None,
+        ))
+        assert_equal(res, None)
 
-        # create new contract instance which can be used to interact
-        # with the instantiated contract
+        # valid data
+        value -= 1
+
+        (tx, (_, blk, _)) = submit_pp_tx(client, tx, alice, value, utxo.Output(
+            value=1,
+            destination=utxo.DestCreatePP(
+                code=os.path.join(os.path.dirname(__file__), "assets/pooltester.wasm"),
+                data=[0xed, 0x4b, 0x9d, 0x1b],
```

Comment:
What is this data?

Reply:
I see it everywhere

Reply:
Selector (like function ID) for the constructor so the Wasm VM knows which function to call when it's instantiating the smart contract. It was a hard lesson to learn that you must always provide a selector..

Reply:
These are Selectors pointing to pooltester.json file (i.e "selector": "0xed4b9d1b")